packages/@lwc/ssr-compiler/src/compile-js/decorators/api/validate.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { DecoratorErrors } from '@lwc/errors';
+import { DISALLOWED_PROP_SET, AMBIGUOUS_PROP_SET } from '@lwc/shared';
+import { generateError } from '../../errors';
+import { type ComponentMetaState } from '../../types';
+import type { Identifier, MethodDefinition, PropertyDefinition } from 'estree';
+
+export type ApiMethodDefinition = MethodDefinition & {
+    key: Identifier;
+};
+export type ApiPropertyDefinition = PropertyDefinition & {
+    key: Identifier;
+};
+
+export type ApiDefinition = ApiPropertyDefinition | ApiMethodDefinition;
+
+function validateName(definition: ApiDefinition) {
+    if (definition.computed) {
+        throw generateError(DecoratorErrors.PROPERTY_CANNOT_BE_COMPUTED);
+    }
+
+    const propertyName = definition.key.name;
+
+    switch (true) {
+        case propertyName === 'part':
+            throw generateError(DecoratorErrors.PROPERTY_NAME_PART_IS_RESERVED, propertyName);
+        case propertyName.startsWith('on'):
+            throw generateError(DecoratorErrors.PROPERTY_NAME_CANNOT_START_WITH_ON, propertyName);
+        case propertyName.startsWith('data') && propertyName.length > 4:
+            throw generateError(DecoratorErrors.PROPERTY_NAME_CANNOT_START_WITH_DATA, propertyName);
+        case DISALLOWED_PROP_SET.has(propertyName):
+            throw generateError(DecoratorErrors.PROPERTY_NAME_IS_RESERVED, propertyName);
+        case AMBIGUOUS_PROP_SET.has(propertyName):
+            throw generateError(
+                DecoratorErrors.PROPERTY_NAME_IS_AMBIGUOUS,
+                propertyName,
+                AMBIGUOUS_PROP_SET.get(propertyName)!
+            );
+    }
+}
```

Comment:
This whole file mirrors the structure and logic from `babel-plugin-component`:

https://github.com/salesforce/lwc/blob/b2468c95b4f187720a9afcc71403e1bf440b7d78/packages/%40lwc/babel-plugin-component/src/decorators/api/validate.ts#L61-L70